File: ggpo/src/lib/GameRecorder.cpp
#include "GameRecorder.h"

#include <vector>
#include <filesystem>
#include <stdexcept>
#include <sstream>


using namespace std;

namespace SessionService
{
  constexpr uint64_t TEST_SESSION_ID = 0;
}

namespace ReplayFile
{
  const vector<uint8_t> Preamble = { 'f', 's', 'n', 'e', 'o', '-', 'r', 'f' };
  extern const vector<uint8_t> Footer = { 'r', 'r', 'x', '-' };
}

namespace StringTools
{
  string Truncate(const string& value, size_t maxLen)
  {
    if (value.size() <= maxLen)
    {
      return value;
    }

    return value.substr(0, maxLen);
  }
}

// ----------------------------------------------------------------------------------------------------
EZWriterEx::EZWriterEx(ostream* toStream_) {
  _Stream = toStream_;
}

// ----------------------------------------------------------------------------------------------------
void EZWriterEx::Write(const uint8_t data) {
  _Stream->write(reinterpret_cast<const char*>(data), static_cast<streamsize>(1));
}

// ----------------------------------------------------------------------------------------------------
template <typename T>
void Write(ostream& stream, const T& value) {
  _Stream->write(reinterpret_cast<const char*>(&value), sizeof(T));
}


namespace EZWriter
{
  template <typename T>
  void Write(ostream& stream, const T& value)
  {
    stream.write(reinterpret_cast<const char*>(&value), sizeof(T));
  }

  void Write(ostream& stream, const uint8_t data)
  {
    stream.write(reinterpret_cast<const char*>(data), static_cast<streamsize>(1));
  }

  void Write(ostream& stream, const vector<uint8_t>& bytes)
  {
    stream.write(reinterpret_cast<const char*>(bytes.data()), static_cast<streamsize>(bytes.size()));
  }

  //void WriteBytes(ostream& stream, const vector<uint8_t>& bytes)
  //{
  //  stream.write(reinterpret_cast<const char*>(bytes.data()), static_cast<streamsize>(bytes.size()));
  //}

  void RawString(ostream& stream, const string& value)
  {
    stream.write(value.data(), static_cast<streamsize>(value.size()));
  }
}

// ------------------------------------------------------------------------------------------------------------------------------
GameRecorder::GameRecorder(const CGameData& gameData_, const string& dataDir, uint64_t sessionId, bool overwriteExisting)
{
  GameData = gameData_;

  if (sessionId == SessionService::TEST_SESSION_ID)
  {
    // TODO: Some kind of passed in or better logger?
    // Log::Debug("Magic session id was used, replay overwrite is enabled!");
    overwriteExisting = true;
  }

  filesystem::path path = filesystem::path(dataDir) / (to_string(sessionId) + ".replay");

  Init(path, overwriteExisting);
}

// ------------------------------------------------------------------------------------------------------------------------------
GameRecorder::GameRecorder(const CGameData& gameData_, const string& toPath, bool overwriteExisting)
{
  GameData = gameData_;
  Init(toPath, overwriteExisting);
}

// ----------------------------------------------------------------------------------------------------------
GameRecorder::~GameRecorder()
{
  CloseStream();
}

// ----------------------------------------------------------------------------------------------------------
void GameRecorder::Init(const filesystem::path& toPath, bool overwriteExisting)
{
  if (filesystem::exists(toPath) && !overwriteExisting)
  {
    throw runtime_error("Data file for session id already exists!");
  }

  
  // int x = gameData.DataSize;

  CreateStream(toPath.string());
  FilePath = toPath.string();
  BaseFrames.fill(-1);
}


// ----------------------------------------------------------------------------------------------------------
bool GameRecorder::HasError() const
{
  return ErrorReason != EErrorReason::None;
}

// ----------------------------------------------------------------------------------------------------------
void GameRecorder::Flush()
{
  if (DataStream.is_open())
  {
    DataStream.flush();
  }
}

// ----------------------------------------------------------------------------------------------------------
void GameRecorder::CompleteReplay(  int frame,  ECompletionReason reason,  EErrorReason errReason,  const string& message)
{
  CheckComplete();

  const int COMPLETE_MSG_LEN = 64;
  // string useMsg = StringTools::Truncate(message, COMPLETE_MSG_LEN);

  stringstream ms(ios::in | ios::out | ios::binary);

  EZWriter::Write(ms, static_cast<uint8_t>(reason));
  EZWriter::Write(ms, static_cast<uint8_t>(errReason));
  EZWriter::Write(ms, frame);

  CopyFixedString(message, COMPLETE_MSG_LEN, WriteBuffer.data(), 0);
  ms.write(reinterpret_cast<const char*>(WriteBuffer.data()), COMPLETE_MSG_LEN);

  EZWriter::Write(ms, ReplayFile::Footer);

  WriteSegmentData(EDataSegmentType::Complete, ms);

  int64_t finalSize = static_cast<int64_t>(DataStream.tellp()) + static_cast<int64_t>(sizeof(int64_t));

  EZWriter::Write(DataStream, finalSize);

  CloseStream();

  RecordingComplete = true;
}

void GameRecorder::CompleteReplay(
  int frame,
  ECompletionReason reason,
  EErrorReason errReason,
  const char* message)
{
  CompleteReplay(frame, reason, errReason, message == nullptr ? string() : string(message));
}

void GameRecorder::AddChatSegment(ChatData& chat)
{
  CheckComplete();

  if (chat.Message.empty())
  {
    return;
  }

  chat.Message = StringTools::Truncate(chat.Message, ChatData::CHAT_DATA_MAX);

  int segmentSize = static_cast<int>(chat.Message.size()) + sizeof(int) + sizeof(int);

  streampos start = DataStream.tellp();

  EZWriter::Write(DataStream, static_cast<uint8_t>(EDataSegmentType::ChatData));
  EZWriter::Write(DataStream, static_cast<uint16_t>(segmentSize));
  EZWriter::Write(DataStream, chat.FromPlayerIndex);
  EZWriter::Write(DataStream, chat.ToPlayerIndex);
  EZWriter::RawString(DataStream, chat.Message);

  streampos end = DataStream.tellp();
  int64_t total = static_cast<int64_t>(end - start);
  int expected = segmentSize + 3;

  if (total != expected)
  {
    throw runtime_error("Data size mismatch on write!");
  }

  DataStream.flush();
}

bool GameRecorder::AddInput(int playerIndex, GameInput& input)
{
  // EZQ<GameInput>& buf = *PlayerBuffers[playerIndex];
  auto& buf = PlayerBuffers[playerIndex];

  if (buf.IsFull())
  {
    OnError(EErrorReason::InputBufferFull, "Too many unmerged inputs!: " + to_string(playerIndex));
    return false;
  }

  int startFrame = BaseFrames[playerIndex];

  if (input.frame == startFrame)
  {
    return false;
  }

  if (input.frame != startFrame + 1)
  {
    throw runtime_error("Invalid frame number!");
  }

  buf.Push(input);
  startFrame++;
  BaseFrames[playerIndex] = startFrame;

  int len = MAX_PLAYERS; // static_cast<int>(PlayerBuffers.size());

  while (true)
  {
    bool popIt = true;
    int startMergeFrame = SyncedBaseFrame;

    for (int i = 0; i < len; i++)
    {
      // EZQ<GameInput>& pBuf = *PlayerBuffers[i];
      auto& pBuf = PlayerBuffers[i];

      if (pBuf.Count() == 0)
      {
        popIt = false;
        break;
      }

      GameInput giBuf;
      pBuf.First(giBuf);

      if (giBuf.frame != startMergeFrame)
      {
        throw runtime_error("Invalid frame number at player index!");
      }
    }

    if (!popIt)
    {
      break;
    }

    for (int i = 0; i < len; i++)
    {
      GameInput giBuf;
      PlayerBuffers[i].First(giBuf);
      MergeBuffer[i] = giBuf;

      PlayerBuffers[i].Pop();
    }

    MergeInputs();

    startMergeFrame++;
    SyncedBaseFrame = startMergeFrame;
  }

  return true;
}

void GameRecorder::CloseStream()
{
  if (DataStream.is_open())
  {
    DataStream.close();
  }
}

void GameRecorder::CreateStream(const string& path)
{
  DataStream.open(path, ios::binary | ios::out | ios::trunc);

  if (!DataStream.is_open())
  {
    throw runtime_error("Unable to create replay data stream.");
  }

  WriteHeader(DataStream);
}

void GameRecorder::WriteHeader(ostream& res)
{
  EZWriter::Write(res, ReplayFile::Preamble);
  EZWriter::Write(res, vector<uint8_t> { 1 });

  WriteGameDataSegment(GameData);
}

void GameRecorder::CheckComplete()
{
  if (RecordingComplete)
  {
    throw runtime_error("Recording is complete! Can't write anymore!");
  }
}

void GameRecorder::WriteInputSegment(GameInput& input)
{
  CheckComplete();

  streampos start = DataStream.tellp();

  int inputSize = GameData.TotalInputSize;
  int segmentSize = inputSize + sizeof(int);

  EZWriter::Write(DataStream, static_cast<uint8_t>(EDataSegmentType::InputData));
  EZWriter::Write(DataStream, static_cast<uint16_t>(inputSize));
  EZWriter::Write(DataStream, input.frame);

  for (int i = 0; i < inputSize; i++)
  {
    WriteBuffer[i] = input.bits[i];
  }

  DataStream.write(reinterpret_cast<const char*>(WriteBuffer.data()), inputSize);

  streampos end = DataStream.tellp();
  int64_t total = static_cast<int64_t>(end - start);
  int expected = segmentSize + 3;

  if (total != expected)
  {
    throw runtime_error("Data size mismatch on write!");
  }

  Flush();
}

void GameRecorder::WriteGameDataSegment(const CGameData& gameData)
{
  CheckComplete();

  streampos start = DataStream.tellp();

  uint16_t segmentSize = CGameData::DataSize;

  EZWriter::Write(DataStream, static_cast<uint8_t>(EDataSegmentType::GameData));
  EZWriter::Write(DataStream, segmentSize);

  {
    int size = CopyFixedString(gameData.GameName, CGameData::MAX_GAME_NAME_SIZE, WriteBuffer.data(), 0);
    DataStream.write(reinterpret_cast<const char*>(WriteBuffer.data()), size);
  }

  {
    int size = CopyFixedString(gameData.GameVersion, CGameData::MAX_VERSION_SIZE, WriteBuffer.data(), 0);
    DataStream.write(reinterpret_cast<const char*>(WriteBuffer.data()), size);
  }

  EZWriter::Write(DataStream, gameData.PlayerCount);
  EZWriter::Write(DataStream, gameData.TotalInputSize);

  streampos end = DataStream.tellp();
  int64_t total = static_cast<int64_t>(end - start);
  int expected = segmentSize + 3;

  if (total != expected)
  {
    throw runtime_error("Data size mismatch on write!");
  }
}

// ------------------------------------------------------------------------------------------------------
void GameRecorder::WriteSegmentData(EDataSegmentType segmentType, stringstream& data)
{
  CheckComplete();

  streampos start = DataStream.tellp();

  data.seekp(0, ios::end);
  int segmentSize = static_cast<int>(data.tellp());

  EZWriter::Write(DataStream, static_cast<uint8_t>(segmentType));
  EZWriter::Write(DataStream, static_cast<uint16_t>(segmentSize));

  data.seekg(0, ios::beg);
  DataStream << data.rdbuf();

  streampos end = DataStream.tellp();
  int64_t total = static_cast<int64_t>(end - start);
  int expected = segmentSize + 3;

  if (total != expected)
  {
    throw runtime_error("Data size mismatch on write!");
  }

  Flush();
}

// --------------------------------------------------------------------------------------------------
int GameRecorder::CopyFixedString(const string& data, int maxSize, uint8_t* toBuffer, int offset)
{
  if (static_cast<int>(data.size()) > maxSize)
  {
    throw runtime_error("Fixed string exceeds maximum size.");
  }

  int len = static_cast<int>(data.size());
  int extra = (std::max)(maxSize - len, 0);

  for (int i = 0; i < len; i++)
  {
    toBuffer[offset + i] = static_cast<uint8_t>(data[i]);
  }

  for (int i = 0; i < extra; i++)
  {
    toBuffer[offset + len + i] = 0;
  }

  return maxSize;
}

void GameRecorder::OnError(EErrorReason errReason, const string& message)
{
  ErrorReason = errReason;
  ErrorMessage = message;

  CompleteReplay(SyncedBaseFrame, ECompletionReason::Error, errReason, message);
}

void GameRecorder::MergeInputs()
{
  // PlayerBuffers.count() == MAX_PLAYERS

  int len = static_cast<int>(MAX_PLAYERS);
  int offset = GameData.TotalInputSize / len;

  GameInput merged{};
  merged.size = GameData.TotalInputSize;
  merged.frame = MergeBuffer[0].frame;

  for (int i = 0; i < len; i++)
  {
    if (MergeBuffer[i].frame != merged.frame)
    {
      throw runtime_error("Unexpected frame number from merge buffer.");
    }

    for (int j = 0; j < offset; j++)
    {
      uint8_t d = MergeBuffer[i].bits[j];
      merged.bits[(i * offset) + j] = d;
    }
  }

  WriteInputSegment(merged);

  MergedInputs.push(merged);
}

File: ReplayAppliance/ReplayAppliance/Clients/GameRecorder.cs
using drewCo.Tools;
using drewCo.Tools.Logging;
using System.Runtime.CompilerServices;
using System.Text;

namespace funscrew
{
  // ==============================================================================================================================
  /// <summary>
  /// This is the thing that records inputs, etc. for a game.
  /// Thre recorded files can then be used for replay, and in some cases, direct streaming.
  /// </summary>
  public class GameRecorder : IDisposable
  {
    private UInt64 SessionId = 0;
    private string DataDir = null!;

    // I need a window of currently active inputs.  We don't want to have them always hanging around.
    // this window of active inputs is also what can be used in cases where we want to stream the game out to
    // active clients.
    // TODO: I think that we can convert this to EZQ as well!
    private RingBuffer<GameInput> MergedInputs = new RingBuffer<GameInput>(64);

    // Each player will also have their inputs in their own slightly larger buffers.
    // These buffers should never overflow and if they so then there is a problem.
    // OPTIONS: Ten seconds @ 60fps: NOTE: This is maybe a bit too big in actual application....
    const int PLAYER_INPUT_BUFFER_SIZE = 60 * 10;     //  0x70;

    // OPTIONS:
    const int MAX_PLAYERS = 2;

    /// <summary>
    /// Used to track what should be the expected starting frame on merge operations when a given buffer is empty!
    /// </summary>
    private int SyncedBaseFrame = 1;

    private EZQ<GameInput>[] PlayerBuffers = null!;
    private int[] BaseFrames = null!;

    private GameInput[] MergeBuffer = null;

    private CGameData GameData = default;

    private byte[] WriteBuffer = new byte[0x800];

    public bool RecordingComplete { get; private set; } = false;
    public bool HasError { get { return ErrorReason != EErrorReason.None; } }
    public string? ErrorMessage { get; private set; } = null!;
    public EErrorReason ErrorReason { get; private set; }


    private ReplayFile ReplayFile = null!;
    public string FilePath { get { return ReplayFile.Path; } }

    // -----------------------------------------------------------------------------------------------------------------------
    // NOTE: In production environments, game data should not be allowed to be overwritten!
    public unsafe GameRecorder(CGameData gameData_, string dataDir_, UInt64 sessionId_, string stateFileName, bool overwriteExisting = false)
    {
      GameData = gameData_;
      SessionId = sessionId_;
      DataDir = dataDir_;

      if (SessionId == GGPOConsts.TEST_SESSION_ID)
      {
        Log.Debug($"Magic session id: {GGPOConsts.TEST_SESSION_ID} was used, replay overwrite is enabled!");
        overwriteExisting = true;
      }

      // Let's create the file.  If it already exists, then we have a problem / invalid session ID!
      string usePath = Path.Combine(DataDir, SessionId + ".replay");
      if (File.Exists(usePath) && !overwriteExisting)
      {
        throw new InvalidOperationException($"Data file for session id: {SessionId} already exists!");
      }

      // NOTE: The emulator will run a lookup for the state file name.
      // There are some weirdo rules / names for ranked / unranked even tho they don't appear to matter...
      // Therefore, I am not going to include any directory information at this time....
      // NOTE: There are a small handful of 'ranked' games like KOF97 that originall have both a '_fbneo' and a '_fbneo_ranked' version.
      // I have no clue what the difference is, or how we might reliably detect it...
      // Probably some kind of bullshit hard-coded lookup table?  Not really sure what to do about it long term....
      // Maybe a game distribution system can figure it out?
      var stateFilePathData = Encoding.UTF8.GetBytes(stateFileName);

      fixed (byte* buffer = stateFilePathData)
      {
        CGameState state = new CGameState();
        state.Type = (uint8_t)EGameStateType.GAMESTATE_TYPE_FILE;
        state.StartFrame = 0;   // 0 == NOT SET.
        state.DataSize = (uint32_t)stateFilePathData.Length;
        state.Data = buffer;

        // NOTE: This would be a CRC32 for the contents of the indicated file!  We use zero for now b/c I don't care about checking it.
        state.CRC32 = 0;

        // We don't need state or its memory after it is written....
        ReplayFile = new ReplayFile(usePath, GameData, state);
        stateFilePathData = null;
      }




      BaseFrames = new int[MAX_PLAYERS];
      int len = BaseFrames.Length;
      for (int i = 0; i < len; i++)
      {
        BaseFrames[i] = 0;
      }
      PlayerBuffers = new EZQ<GameInput>[MAX_PLAYERS];
      for (int i = 0; i < MAX_PLAYERS; i++)
      {
        PlayerBuffers[i] = new EZQ<GameInput>(PLAYER_INPUT_BUFFER_SIZE);
      }
      MergeBuffer = new GameInput[MAX_PLAYERS];
    }

    // -----------------------------------------------------------------------------------------------------------------------
    public void Dispose()
    {
      ReplayFile.Dispose();
    }

    // -----------------------------------------------------------------------------------------------------------------------
    /// <summary>
    /// Use this to complete the recording of the replay + indicate the
    /// reason why it was completed.  This could be through a proper disconnect,
    /// or an error, etc.
    /// </summary>
    public void CompleteReplay(int frame, ECompletionReason reason, EErrorReason errReason, string? message)
    {
      if (RecordingComplete) { return; }
      int useFrame = frame == -1 ? SyncedBaseFrame : frame;
      ReplayFile.CompleteWrite(reason, errReason, string.Empty);
      RecordingComplete = true;
    }

    // -----------------------------------------------------------------------------------------------------------------------
    public void AddChatSegment(ref CChatData chat)
    {
      if (string.IsNullOrWhiteSpace(chat.Message)) { return; }
      chat.Message = StringTools.Truncate(chat.Message, CChatData.CHAT_DATA_MAX);

      ReplayFile.AddChat(ref chat);
    }

    // -----------------------------------------------------------------------------------------------------------------------
    /// <returns>
    /// Boolean value indicating if the frame was accepted.
    /// Returns false if the frame already exists</returns>
    public unsafe bool AddInput(int playerIndex, ref GameInput input)
    {
      var buf = PlayerBuffers[playerIndex];

      if (buf.IsFull)
      {
        // TODO: Shutdown the recorder correctly!
        // Send other events!
        // We don't want to use exceptions for flow control, rather we need to set an error state on the recorder!
        // throw new InvalidOperationException($"Input buffer for player: {playerIndex} is full!");

        this.OnError(EErrorReason.InputBufferFull, $"Too many unmerged inputs from Player:{playerIndex + 1}! (size={buf.Capacity})");
        return false;
      }

      // It is OK if we add a duplicate frame.
      // NOTE: This is a copy we can probably avoid!
      // The start frame should be per-player, otherwise they have to always be in perfect sync...
      int startFrame = BaseFrames[playerIndex]; /// SyncedBaseFrame;
      if (input.frame == startFrame)
      {
        // Ignore duplicate frame.
        // TODO: Log this, we may not actually need it...
        return false;
      }

      if (input.frame != startFrame + 1)
      {
        // TODO: Close the recording here, properly!
        throw new InvalidOperationException("Invalid frame number!");
      }
      buf.Push(input);
      startFrame++;
      BaseFrames[playerIndex] = startFrame;

      int len = this.PlayerBuffers.Length;
      bool popIt = true;

      while (true)
      {
        int startMergeFrame = SyncedBaseFrame;

        // throw new InvalidOperationException("refigure how to do the sync/global base frames");

        // Now that we have added a frame, we will go to the back, and find
        // all frames that match + do the merge.
        for (int i = 0; i < len; i++)
        {
          var pBuf = this.PlayerBuffers[i];
          if (pBuf.Count == 0)
          {
            popIt = false;
            break;
          }

          GameInput giBuf = new GameInput();
          pBuf.First(ref giBuf);
          if (giBuf.frame != startMergeFrame)
          {
            throw new InvalidOperationException($"Invalid (back) frame number at player index: {i}!  Should be {startMergeFrame}!");
          }
        }

        // Nothing left to confirm!
        if (!popIt) { break; }

        // Copy data for the merge!
        for (int i = 0; i < len; i++)
        {
          var giBuf = new GameInput();
          this.PlayerBuffers[i].First(ref giBuf);
          MergeBuffer[i] = giBuf;

          this.PlayerBuffers[i].Pop();
        }

        MergeInputs();


        // Do the merge here + write the segment!
        startMergeFrame++;
        SyncedBaseFrame = startMergeFrame;
      }

      return true;
    }

    // -----------------------------------------------------------------------------------------------------------------------
    private void OnError(EErrorReason errReason, string message)
    {
      this.ErrorReason = errReason;
      this.ErrorMessage = message;
      CompleteReplay(this.SyncedBaseFrame, ECompletionReason.Error, errReason, message);
    }

    // -----------------------------------------------------------------------------------------------------------------------
    private unsafe void MergeInputs()
    {
      // We want to create a single GameInput instance from a full set.
      int len = this.PlayerBuffers.Length;
      int offset = this.GameData.TotalInputSize / len;

      // Reminder:  The original game inputs all come with the data in the P0 slot.
      // Therefore we need to be make sure that they end up in their game-correct
      // positions for playback / correct interpretation later.
      GameInput merged = new GameInput();
      merged.size = this.GameData.TotalInputSize;
      merged.frame = MergeBuffer[0].frame;

      for (int i = 0; i < len; i++)
      {
        // Make sure that the frames are correct (NOTE: This may be done in a previous step already....);
        if (MergeBuffer[i].frame != merged.frame)
        {
          // CompleteReplay(
          throw new InvalidOperationException($"Unexpected frame # from merge buffer: {i}! ({merged.frame} {MergeBuffer[i].frame})");
        }
        for (int j = 0; j < offset; j++)
        {
          byte d = MergeBuffer[i].data[j];
          merged.data[(i * offset) + j] = d;
        }
      }

      // Write that data to disk!
      ReplayFile.AddInput(ref merged);

      // Add it to the active window of inputs (which are used for live playback)
      this.MergedInputs.Push(merged);
    }
  }

}
